### ISSUE TITLE
Optimize cursor movement

### ISSUE THREAD

**Issue by raineorshine** [cybersemics/em#484]
From https://github.com/cybersemics/em/pull/421#issuecomment-605366333:

> FYI, performance is at least no worse than the existing implementation. Tested with ~10k thoughts. They're both bad. Looking forward to optimizing in a separate task.
> 
> `dev`: 
> cursorUp:       260ms
> cursorDown:   802ms
> cursupUp:      312ms
> cursorDown:   706ms
> cursorUp:       280ms
> cursorDown:   825ms
> cursupUp:      247ms
> cursorDown:   730ms
> cursupUp:      344ms
> 
> `restore-selection` (PR)
> cursorUp:       250ms
> cursorDown:   500ms
> cursupUp:      352ms
> cursorDown:   521ms
> cursorUp:       750ms
> cursorDown:   200ms + 200ms (keyup and keydown)
> cursupUp:      600ms
> cursorDown:   223ms + 165ms (keyup and keydown)
> cursupUp:      522ms

I have not yet measured how much is the result of business logic and how much is rendering. #188 is the first major step towards render optimization.

**Comment by raineorshine** [comment]
Rendering performance is still quite still, but let's test after #188.
